cub/test/catch2_test_device_merge_sort.cu

```diff
@@ -0,0 +1,309 @@
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/random.h>
+#include <thrust/sequence.h>
+#include <thrust/shuffle.h>
+
+#include <algorithm>
+
+#include "catch2_test_device_merge_sort_common.cuh"
+#include "catch2_test_helper.h"
+#include "catch2_test_launch_helper.h"
+
+// %PARAM% TEST_LAUNCH lid 0:1:2
+
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::SortPairs, sort_pairs);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::SortPairsCopy, sort_pairs_copy);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::StableSortPairs, stable_sort_pairs);
+
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::SortKeys, sort_keys);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::SortKeysCopy, sort_keys_copy);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::StableSortKeys, stable_sort_keys);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceMergeSort::StableSortKeysCopy, stable_sort_keys_copy);
+
+using key_types =
+  c2h::type_list<std::uint8_t,
+                 std::int16_t,
+                 std::uint32_t,
+                 double,
+                 c2h::custom_type_t<c2h::equal_comparable_t, c2h::less_comparable_t>>;
+using wide_key_types = c2h::type_list<std::uint32_t, double>;
+
+using value_types =
+  c2h::type_list<std::uint8_t, float, c2h::custom_type_t<c2h::equal_comparable_t, c2h::less_comparable_t>>;
+
+/**
+ * Function object that maps the targeted sorted rank of an item to a key.
+
+ * E.g., `OffsetT` is `int32_t` and `KeyT` is `float`:
+ * [  4,   2,   3,   1,   0] <= targeted key ranks
+ * [4.0, 2.0, 3.0, 1.0, 0.0] <= corresponding keys
+ */
+template <typename OffsetT, typename KeyT>
+struct rank_to_key_op_t
+{
+  __device__ __host__ KeyT operator()(const OffsetT& val)
+  {
+    return static_cast<KeyT>(val);
+  }
+};
+
+template <typename OffsetT>
+struct rank_to_key_op_t<OffsetT, c2h::custom_type_t<c2h::equal_comparable_t, c2h::less_comparable_t>>
+{
+  using custom_t = c2h::custom_type_t<c2h::equal_comparable_t, c2h::less_comparable_t>;
+  __device__ __host__ custom_t operator()(const OffsetT& val)
+  {
+    custom_t custom_val{};
+    custom_val.key = val;
+    custom_val.val = val;
+    return custom_val;
+  }
+};
+
+/**
+ * Generates a shuffled array of key ranks. E.g., for a vector of size 5: [4, 2, 3, 1, 0]
+ */
+template <typename OffsetT>
+thrust::device_vector<OffsetT> make_shuffled_key_ranks_vector(OffsetT num_items, c2h::seed_t seed)
+{
+  thrust::device_vector<OffsetT> key_ranks(num_items);
+  thrust::sequence(key_ranks.begin(), key_ranks.end());
+  thrust::shuffle(
+    key_ranks.begin(), key_ranks.end(), thrust::default_random_engine{static_cast<unsigned int>(seed.get())});
+  return key_ranks;
+}
+
+CUB_TEST("DeviceMergeSort::SortKeysCopy works", "[merge][sort][device]", wide_key_types)
+{
+  using key_t    = typename c2h::get<0, TestType>;
+  using offset_t = std::int32_t;
+
+  // Prepare input
+  // const offset_t num_items = GENERATE_COPY(take(2, random(1, 1000000)), values({500, 1000000, 2000000}));
```

Comment:
question: is this a leftover from debugging? I guess we want this line uncommented and the successor line removed? If this was intended, I'd just initialize num items without using GENERATE.

Reply:
Good catch, thanks, that was indeed an artefact from verifying that we would correctly generate test data.